### ISSUE TITLE
Update Docker Image / Optimize Runner Update

### ISSUE THREAD

**Issue by MrSerth** [openHPI/poseidon#33]
As part of #22, we discussed optimization strategies for Runner updates. One idea was to skip updating runners if nothing changed (which in general is a good idea). For most values (e.g. the RAM oder CPU shares), this would work and reduce the load on Nomad. One part we weren't sure about was updating the Docker image.

In my tests, I found out that the Docker image will not be updated automatically (in Nomad) when updating the runner though Poseidon. According to the Nomad API, images with a specific tag (i.e. not `latest`) are not updated automatically. If all images should be updated, the [`force_pull`](https://www.nomadproject.io/docs/drivers/docker#force_pull) option in the task configuration could be used.

This option `force_pull = true` would be added in the config of the job:
https://github.com/openHPI/poseidon/blob/fff67246d6cc0bad92f403f55dd77dc5bcf62b6c/internal/environment/template-environment-job.hcl#L31-L36

When doing so, Nomad will pull the latest image for the tag provided and recreate environments when the image has changed. This is probably something we want to implement; I tested it successfully on my local machine.

However, when we rely on this Nomad feature, we probably cannot optimize the update process by skipping runner updates if nothing changed (the image could have been updated under the same name and thus no change is visible). A possible optimization could be to enable [rolling updates](https://learn.hashicorp.com/tutorials/nomad/job-rolling-update) which seem quite easy and would reduce the workload we put on Nomad when updating an environment with a large pool size.

**Comment by MrSerth** [comment]
As discussed, we proposed to `force_pull` images only when updating the execution environment and not when creating new runners.

**Comment by MrSerth** [comment]
A workaround, for now, is the following:

1. Reduce the pool size to 0 in CodeOcean
2. Sync all environments and pay attention that all allocations are removed
3. Update the template job manually by injecting the `force_pull` option
4. Increase the pool size back to normal in CodeOcean

This will ensure that all runners get recreated with the updated image.

**Comment by mpass99** [comment]
Rolling updates are not allowed with "batch" jobs. But we use this type of jobs. Judging from [this documentation](https://www.nomadproject.io/docs/schedulers) we should not change our current job type. Therefore I would refrain from the rolling update feature.

**Comment by mpass99** [comment]
> As discussed, we proposed to `force_pull` images only when updating the execution environment and not when creating new runners.

Might this lead to errors? Example: All runners of the execution environment are distributed on the Nomad Workers except on Nomad Worker 4. Now the runners and thus the Docker image are updated on the other workers. But when the next runner is scheduled on Worker 4, it has the old image again?!
To avoid this, we can set the flag permanently, as you suggested first?!

**Comment by MrSerth** [comment]
> Rolling updates are not allowed with "batch" jobs. But we use this type of jobs. Judging from [this documentation](https://www.nomadproject.io/docs/schedulers) we should not change our current job type. Therefore I would refrain from the rolling update feature.

Okay, I am convinced. We drop the idea of rolling updates.

> ...
> To avoid this, we can set the flag permanently, as you suggested first?!

Sure, this might work and is a workaround. Still, we might hit the pull limit and potentially slow down the creation of new allocations. When I performed my manual update last week, I thought that stopping all allocations, updating the template job and recreating the other allocations would do the job. I thought this mechanism would work because of the [automated Image cleanup](https://www.nomadproject.io/docs/drivers/docker#docker-cleanup-image). I just learned it has a timeout of 3 minutes - and I probably was faster than that. Maybe we could change that to force an update?

If there is no great solution, I could also introduce explicit versioning support to the images (with changing image tags). This, for sure, should work... 🙈

**Comment by mpass99** [comment]
> Sure, this might work and is a workaround. Still, we might hit the pull limit [...]

I think the pull limit is for the downloads and not for just comparing the digests. At least I just could pull over 350 times the same image without running into this issue. 

> [...] and potentially slow down the creation of new allocations.

How big would the difference be to the current state? `force_pull` is activated for all images without tag and with the `:lastest` tag by Nomads default. Aren't these most of the images?

> Maybe we could change that to force an update?

Yes. We can lower the timeout for the automated image cleanup if you are ok with the process of manually lowering and raising the pool size for this use case.

**Comment by MrSerth** [comment]
With [my latest change to the Dockerfiles](https://github.com/openHPI/dockerfiles/commit/a4ec9e58569e037f113880210b8c01f5b6625ee3), I dropped the `:latest` tag and no longer use it for any image (in favor of preventing issues with updates to existing images that are not backwards compatible). For example, we previously had Python 3.4 and 3.8 where 3.4 used the latest tag...

```
REPOSITORY                  TAG           IMAGE ID       CREATED       SIZE
openhpi/co_execenv_r        4             c80c1a026481   7 days ago    668MB
openhpi/co_execenv_python   3.4-rpi-web   a712445639ff   3 weeks ago   579MB
openhpi/co_execenv_python   3.4           eb35c9cc95a6   3 weeks ago   353MB
openhpi/co_execenv_node     0.12          f13dc296c84d   3 weeks ago   437MB
openhpi/co_execenv_java     17            724e17c0aa40   3 weeks ago   837MB
openhpi/co_execenv_ruby     2.5           015bb79ca511   3 weeks ago   361MB
openhpi/co_execenv_python   3.7-ml        fe94670dda8e   3 weeks ago   1.28GB
openhpi/co_execenv_python   3.8           24a4ae3e74c7   3 weeks ago   336MB
openhpi/co_execenv_java     8-antlr       672379cee2ba   3 weeks ago   698MB
```

I also thought that this would benefit the Nomad cluster in high load scenarios due to the reduced load involved in _just_ reusing the current image. I was thinking of the frequency we create new allocations as the unused runner time is currently set to three minutes.

In the latest test I just did, I was able to confirm that the proposed mechanism (set pool size to 0, update template, restore pool size) works *if* the images are auto-cleaned in-between. Then, the allocations will first redownload the image and finally start the runners. A potential side effect of this mechanism would be that images get updated / downloaded more often, depending on Nomad's scheduler.

I am slo fine with sending a flag in the PUT request to Poseidon. This way, we could either reuse the existing route or create a new one and only include the `force_pull` option if this flag is true.
